flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/TimestampsAndWatermarksOperatorTest.java

```diff
@@ -87,6 +88,18 @@ public void periodicWatermarksEmitOnPeriodicEmit() throws Exception {
 
 		assertThat(pollNextStreamRecord(testHarness), streamRecord(4L, 4L));
 		assertThat(pollNextLegacyWatermark(testHarness), is(legacyWatermark(3L)));
+
+		testHarness = createBatchHarness(WatermarkStrategy.forGenerator((ctx) -> new PeriodicWatermarkGenerator())
```

Comment:
Let's extract it to a separate test. It's always better to test a single case at a time. I would rename the existing test to:
`periodicWatermarksEmitOnPeriodicEmitStreamMode` and the batch one `periodicWatermarksEmitOnPeriodicEmitBatchMode`. The same for the other test.

Moreover in the batch case you are not advancing the processing time. Therefore you are not testing the periodic generator. It passes for `STREAM` mode as well.

Lastly could you add tests for the punctuated watermark generator? The `PeriodicWatermarkGenerator` emits only in `onPeriodicEmit`, we should test that we do not emit from `onEvent` as well.